keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md

```diff
@@ -0,0 +1,941 @@
++ 		return true, nil, nil
+	}
+
+	// Look for a matching CSIStorageCapacity object(s).
+	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure (https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718).
+	capacities, err := b.csiStorageCapacityLister.List(labels.Everything())
+	if err != nil {
+- 		return false, err
++ 		return false, nil, err
+	}
+
+  sizeInBytes := quantity.Value()
+	for _, capacity := range capacities {
+		if capacity.StorageClassName == storageClass.Name &&
+			capacitySufficient(capacity, sizeInBytes) &&
+			b.nodeHasAccess(node, capacity) {
+			// Enough capacity found.
+- 			return true, nil
++ 			return true, capacity, nil
+		}
+	}
+
+	// TODO (?): this doesn't give any information about which pools where considered and why
+	// they had to be rejected. Log that above? But that might be a lot of log output...
+	klog.V(4).InfoS("Node has no accessible CSIStorageCapacity with enough capacity for PVC",
+		"node", klog.KObj(node), "PVC", klog.KObj(claim), "size", sizeInBytes, "storageClass", klog.KObj(storageClass))
+- 	return false, nil
++ 	return false, nil, nil
+}
+```
+
+### Scoring of nodes for dynamic provisioning
+
+Add scoring of nodes for dynamic provisioning in the `Score` method of the VolumeBinding plugin. The scoring applies to every entry in `podVolumes.DynamicProvisions` where `Capacity` is not equal to `nil`.
+
+Scoring is implemented using the existing StaticBindings mechanism. Add the following to `classResources`, which will be passed to the `scorer` function:
+
+- `Requested: provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]`
+- `Capacity: CSIStorageCapacity`
+
+By doing this, we can assign scores to nodes for dynamic provisioning in a way that is based on the `Shape` setting of `VolumeBindingArgs`, and which takes into account the amount of free space the nodes have.
+
+```diff
+// Score invoked at the score extension point.
+func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	if pl.scorer == nil {
+		return 0, nil
+	}
+	state, err := getStateData(cs)
+	if err != nil {
+		return 0, framework.AsStatus(err)
+	}
+	podVolumes, ok := state.podVolumesByNode[nodeName]
+
+	...
+
++ 	// add dynamic binding volumes by storage class
++ 	for _, provision := range podVolumes.DynamicProvisions {
++ 		if provision.Capacity == nil {
++ 			continue
++ 		}
++ 		class := *provision.PVC.Spec.StorageClassName
++ 		if _, ok := classResources[class]; !ok {
++ 			classResources[class] = &StorageResource{
++ 				Requested: 0,
++ 				Capacity:  0,
++ 			}
++ 		}
++ 		requestedQty := provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
++ 		classResources[class].Requested += requestedQty.Value()
++ 		classResources[class].Capacity += provision.Capacity.Capacity.Value()
++ 	}
+
+  return pl.scorer(classResources), nil
+```
+
+### Test Plan
```

Comment:
The remaining sections in the kep template should also be filled out.

Reply:
This is a work in progress. It will take a while.

Reply:
@msau42 I filled out these sections. Please review.